### PR TITLE
[Windows] Introduce `egl::Surface` and `egl::WindowSurface`

### PR DESCRIPTION
**Pull request by loic-sharma**
This introduces the `egl::Surface` and `egl::WindowSurface` types to abstract a raw `EGLSurface`. This also removes some - but not all - EGL surface logic out from `EGLManager`.

Subsequent pull requests will be necessary to:

1. Move ownership of the `egl::WindowSurface` from `egl::Manager` to `FlutterWindowsView`
2. Refactor external texture's off-screen EGL surface to use `egl::Surface`

Part of https://github.com/flutter/flutter/issues/141996

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by yaakovschectman — APPROVED** [review]
LGTM

File: shell/platform/windows/egl/context.h [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace egl {
 
 // An EGL context to interact with OpenGL.
 //
-// This enables automatic eror logging and mocking.
+// This enables automatic error logging and mocking.
```
Comment (cbracken) [inline shell/platform/windows/egl/context.h:17]: Nice work catching this eror.

File: shell/platform/windows/egl/surface.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                         EGL_NO_CONTEXT) != EGL_TRUE) {
+      WINDOWS_LOG_EGL_ERROR;
+      return false;
+    }
+
+    if (::eglDestroySurface(display_, surface_) != EGL_TRUE) {
+      WINDOWS_LOG_EGL_ERROR;
+      return false;
+    }
+  }
```
Comment (cbracken) [inline shell/platform/windows/egl/surface.cc:36]: Are we enforcing anywhere which thread this class is called from? If not, should we add debug build FML checks to avoid the potential for race conditions if someone down the road changes things?

e.g. if this were called from two threads and the first thread were just about to start executing the code at line 26 as the second finishes the executing the whole block ending line 36, things would be sad

**Review by cbracken — APPROVED** [review]
lgtm - much cleaner. Reviewing these is a joy.

LGTM stamp from a Japanese personal seal

File: shell/platform/windows/egl/surface.cc (hunk 2 above)
Comment (loic-sharma) [reply shell/platform/windows/egl/surface.cc:36]: We're not enforcing which thread this is called from yet, but we really should given OpenGL's behavior is heavily dependent on the current thread.

There's some surface operations that happen on both the platform and raster threads:

1. Make a surface current: on the platform thread for initialization, on the raster thread otherwise
1. Destroying a surface: on raster thread for resizing, or on platform thread for engine shutdown

In other words, this layer doesn't know enough to assert the current thread is expected. We likely will want to sprinkle these kinds of assertions in the view and compositor abstractions. I'll need to think more on this :)
